go/vt/vtgate/planbuilder/operators/join.go

```diff
@@ -83,38 +85,49 @@ func (j *Join) Compact(ctx *plancontext.PlanningContext) (Operator, *ApplyResult
 	return newOp, Rewrote("merge querygraphs into a single one")
 }
 
-func createOuterJoin(tableExpr *sqlparser.JoinTableExpr, lhs, rhs Operator) Operator {
-	if tableExpr.Join == sqlparser.RightJoinType {
+func createLeftAndStraightJoin(ctx *plancontext.PlanningContext, join *sqlparser.JoinTableExpr, lhs, rhs Operator) Operator {
+	// first we switch sides, so we always deal with left outer joins
+	switch join.Join {
+	case sqlparser.RightJoinType:
 		lhs, rhs = rhs, lhs
+		join.Join = sqlparser.LeftJoinType
+	case sqlparser.NaturalRightJoinType:
+		lhs, rhs = rhs, lhs
+		join.Join = sqlparser.NaturalLeftJoinType
+	}
+
+	// for inner joins we can treat the predicates as filters on top of the join
+	joinOp := &Join{LHS: lhs, RHS: rhs, JoinType: join.Join}
+	if join.Join.IsInner() {
+		return addJoinPredicates(ctx, join.Condition.On, joinOp)
 	}
-	subq, _ := getSubQuery(tableExpr.Condition.On)
+
+	// for outer joins we have to be a bit more careful
+	var op Operator
+	subq, _ := getSubQuery(join.Condition.On)
 	if subq != nil {
 		panic(vterrors.VT12001("subquery in outer join predicate"))
 	}
-	predicate := tableExpr.Condition.On
+	predicate := join.Condition.On
 	sqlparser.RemoveKeyspaceInCol(predicate)
-	return &Join{LHS: lhs, RHS: rhs, LeftJoin: true, Predicate: predicate}
-}
+	joinOp.Predicate = predicate
+	op = joinOp
 
-func createJoin(ctx *plancontext.PlanningContext, LHS, RHS Operator) Operator {
-	lqg, lok := LHS.(*QueryGraph)
-	rqg, rok := RHS.(*QueryGraph)
-	if lok && rok {
-		op := &QueryGraph{
-			Tables:     append(lqg.Tables, rqg.Tables...),
-			innerJoins: append(lqg.innerJoins, rqg.innerJoins...),
-			NoDeps:     ctx.SemTable.AndExpressions(lqg.NoDeps, rqg.NoDeps),
-		}
-		return op
-	}
-	return &Join{LHS: LHS, RHS: RHS}
+	return op
```

Comment:
we should split the method into two as outer join and Straight Join have different code paths.